noether_tpp/src/tool_path_modifiers/waypoint_orientation_modifiers.cpp

```diff
@@ -104,4 +104,89 @@ ToolPaths UniformOrientationModifier::modify(ToolPaths tool_paths) const
   return tool_paths;
 };
 
+static Eigen::Quaterniond computeQuaternionMean(const std::vector<Eigen::Quaterniond>& quaterns)
+{
+  /* Mean quaternion is found using method described by Markley et al: Quaternion Averaging
+   * https://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/20070017872.pdf
+   *
+   * M = sum(w_i * q_i * q_i^T)    Eq. 12
+   * q_bar = argmax(q^T * M * q)   Eq. 13
+   *
+   * "The solution of this maximization problem is well known. The average quaternion is
+   * the eigenvector of M corresponding to the maximum eigenvalue."
+   *
+   * In the above equations, w_i is the weight of the ith quaternion.
+   * In this case, all quaternions are equally weighted (i.e. w_i = 1)
+   */
+
+  Eigen::Matrix4d M = Eigen::Matrix4d::Zero();
+
+  for (const Eigen::Quaterniond& q : quaterns)
+  {
+    M += q.coeffs() * q.coeffs().transpose();
+  }
+
+  // Calculate the SVD of the M matrix
+  Eigen::JacobiSVD<Eigen::Matrix4d> svd(M, Eigen::ComputeFullU);
+
+  // The eigenvectors are represented by the columns of the U matrix; the eigenvector corresponding to the largest
+  // eigenvalue is in row 0
+  Eigen::Quaterniond q;
+  q.coeffs() << svd.matrixU().col(0);
+
+  assert(std::isnan(q.w()) == false && std::isnan(q.x()) == false && std::isnan(q.y()) == false &&
+         std::isnan(q.z()) == false);
+
+  return q;
+};
+
+static Eigen::Quaterniond getMovingAverageQuaternion(const ToolPathSegment& input_poses,
+                                                     const int pose_index,
+                                                     const int n_pts)
+{
+  Eigen::Quaterniond out(input_poses[static_cast<std::size_t>(pose_index)].linear());
+
+  int n = static_cast<int>(input_poses.size());
+  if (pose_index > 0 && pose_index < n - 1)
+  {
+    int n_each_side = n_pts / 2;
+    int start_idx = std::max(pose_index - n_each_side, 0);
+    int stop_idx = std::min(pose_index + n_each_side, n - 1);
+
+    // Accumulate all quaternions
+    std::vector<Eigen::Quaterniond> q;
+    q.reserve(input_poses.size());
+    std::transform(input_poses.begin() + start_idx,
+                   input_poses.begin() + stop_idx + 1,
+                   std::back_inserter(q),
+                   [](const Eigen::Isometry3d& p) { return Eigen::Quaterniond(p.linear()); });
```

Comment:
Is there a computational/resource/speed advantage to using `std::transform` here over a simple `for` loop?  Or are you just revelling in your powers as a std::sorcerer?